capstone/src/main/java/com/google/sps/servlets/UsersServlet.java

```diff
@@ -0,0 +1,294 @@
+   * @return Boolean
+   */
+  public Boolean checkUniversityIdExists(
+    DataSource pool,
+    UserService userService,
+    HttpServletResponse response,
+    HashMap<String, String> resultParameters
+  )
+    throws SQLException, IOException {
+    Integer university_id = Integer.parseInt(
+      resultParameters.get("university")
+    );
+    try (
+      Connection conn = pool.getConnection();
+      PreparedStatement queryStatement = conn.prepareStatement("SELECT * FROM university WHERE university.id = ?")
+    ) {
+      queryStatement.setInt(1, university_id);
+      ResultSet result = queryStatement.executeQuery();
+      if (result.next() == false) {
+        response.sendError(403, "University's id sent by client does not exist.");
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * This method saves the user's information in the request parameters to our user_info table.
+   * @param pool It is a DataSource object that serves to interact with our database connection.
+   * @param userService This is a UserService object that holds information about the current user.
+   * @param resultParameters  A HashMap object that holds all the information sent in the "body" of the user's request.
+   * @return void
+   */
+  public void saveUserDetails(
+    DataSource pool,
+    UserService userService,
+    HashMap<String, String> resultParameters
+  )
+    throws SQLException {
+    try (
+      Connection conn = pool.getConnection();
+      PreparedStatement queryStatement = conn.prepareStatement(
+        CREATE_NEW_USER_QUERY
+      )
+    ) {
+      String user_email = userService.getCurrentUser().getEmail();
+      queryStatement.setString(1, resultParameters.get("user_name"));
+      queryStatement.setString(2, resultParameters.get("full_name"));
+      queryStatement.setString(3, user_email);
+      queryStatement.setInt(
+        4,
+        Integer.parseInt(resultParameters.get("university"))
+      );
+      queryStatement.execute();
+    }
+  }
+
+  /**
+   * This method checks if the POST request is authorized or not.
+   * @param pool It is a DataSource object that serves to interact with our database connection.
+   * @param userService This is a UserService object that holds information about the current user.
+   * @param response A HttpServletResponse object that holds the information that will be sent back to the user.
+   * @return Boolean
+   */
+   public Boolean isPostRequestAuthorized(UserService userService, DataSource pool, HttpServletResponse response) 
+   throws IOException{
+       Boolean isUserLoggedIn = userService.isUserLoggedIn();
+        if (!isUserLoggedIn) {
+            response.sendRedirect("/login");
+            return false;
+        }
+        String email = userService.getCurrentUser().getEmail();
+        String logoutUrl = userService.createLogoutURL(URL_TO_REDIRECT_AFTER_USER_LOGS_OUT);
+        if (!email.endsWith("google.com")) {
```

Comment:
Be very careful with anything security related. Double, triple check it for any edge cases. Ask someone for help if you're unsure. I could go register totally-not-google.com and my email would get past this check. I _think_ `endsWith("@google.com")` is safe but for production code I would ask a security expert. Email address can have all sorts of weird characters: https://en.wikipedia.org/wiki/Email_address#Syntax.